### PR TITLE
added some bfs and topo sort problems.

### PR DESCRIPTION
**Pull request by wodeleeway**

### CODE REVIEW

File: python/bfs_heap_topological_sort/bfs/leetcode_200_number_of_islands_frq100.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+Example 1:
+
+Input: grid = [
+  ["1","1","1","1","0"],
+  ["1","1","0","1","0"],
+  ["1","1","0","0","0"],
+  ["0","0","0","0","0"]
+]
+Output: 1
+Example 2:
```
Comment (zjkang) [inline python/bfs_heap_topological_sort/bfs/leetcode_200_number_of_islands_frq100.py]: format

File: python/bfs_heap_topological_sort/bfs/lintcode_178_graph_valid_tree.medium.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+You can assume that no duplicate edges will appear in edges. Since all edges are undirected, [0, 1] is the same as [1, 0] and thus will not appear together in edges.
+"""
+class Solution:
+    """
+    @param n: An integer
+    @param edges: a list of undirected edges
+    @return: true if it's a valid tree, or false
+    """
+    def validTree(self, n, edges):
+        # write your code here
```
Comment (zjkang) [inline python/bfs_heap_topological_sort/bfs/lintcode_178_graph_valid_tree.medium.py]: minor: remove comment

File: python/bfs_heap_topological_sort/bfs/lintcode_573_build_post_office_ii_hard.py [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+
+Return the smallest sum of distance. Return -1 if it is not possible.
+
+样例
+Example 1:
+
+Input：[[0,1,0,0,0],[1,0,0,2,1],[0,1,0,0,0]]
+Output：8
+Explanation： Placing a post office at (1,1), the distance that post office to all the house sum is smallest.
+Example 2:
```
Comment (zjkang) [inline python/bfs_heap_topological_sort/bfs/lintcode_573_build_post_office_ii_hard.py]: format

File: python/bfs_heap_topological_sort/bfs/lintcode_573_build_post_office_ii_hard.py [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+
+Input：[[0,1,0],[1,0,1],[0,1,0]]
+Output：4
+Explanation： Placing a post office at (1,1), the distance that post office to all the house sum is smallest.
+注意事项
+You cannot pass through wall and house, but can pass through empty.
+You only build post office on an empty.
+
```
Comment (zjkang) [inline python/bfs_heap_topological_sort/bfs/lintcode_573_build_post_office_ii_hard.py:27]: is there any constrain mentioned? about number of houses and empty space?

File: python/bfs_heap_topological_sort/topo/lintcode_892_alien_dictionary_hard.py [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+            node = heapq.heappop(queue)
+            ans += node
+            
+            for neigbor in graph[node]:
+                indegres[neigbor] -= 1
+                
+                if indegres[neigbor] == 0:
+                    heapq.heappush(queue, neigbor)
+        
+        print('ans: ' + str(ans))
```
Comment (zjkang) [inline python/bfs_heap_topological_sort/topo/lintcode_892_alien_dictionary_hard.py]: remove print

File: python/bfs_heap_topological_sort/topo/lintcode_892_alien_dictionary_hard.py [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+        if len(ans) == len(indegres):
+            return ans
+            
+        return ""        
+    
+    def getIndegres(self, graph):
```
Comment (zjkang) [inline python/bfs_heap_topological_sort/topo/lintcode_892_alien_dictionary_hard.py:73]: minor: getIndegrees?

File: python/bfs_heap_topological_sort/topo/lintcode_892_alien_dictionary_hard.py [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+        # add the edges
+        n = len(words)
+        
+        for i in range(n - 1):
+            for j in range(min(len(words[i]), len(words[i + 1]))):
+                if words[i][j] != words[i + 1][j]:
+                    graph[words[i][j]].add(words[i + 1][j])
+                    break
+    
+
```
Comment (zjkang) [inline python/bfs_heap_topological_sort/topo/lintcode_892_alien_dictionary_hard.py:101]: extra line to be removed

File: python/bfs_heap_topological_sort/bfs/lintcode_618_search_graph_nodes_medium.py [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+    @param: target: An integer
+    @return: a node
+    """
+    def searchNode(self, graph, values, node, target):
+        # write your code here
+        
+        queue = collections.deque([node])
+        visited = set([node])
+        
+        closest = None
```
Comment (zjkang) [inline python/bfs_heap_topological_sort/bfs/lintcode_618_search_graph_nodes_medium.py:73]: you not need it, just return None if not found

File: python/bfs_heap_topological_sort/bfs/lintcode_618_search_graph_nodes_medium.py [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+class Solution:
+    """
+    @param: graph: a list of Undirected graph node
+    @param: values: a hash mapping, <UndirectedGraphNode, (int)value>
+    @param: node: an Undirected graph node
+    @param: target: An integer
+    @return: a node
+    """
+    def searchNode(self, graph, values, node, target):
+        # write your code here
```
Comment (zjkang) [inline python/bfs_heap_topological_sort/bfs/lintcode_618_search_graph_nodes_medium.py]: minor: remove

File: python/bfs_heap_topological_sort/bfs/lintcode_598_zombie_in_matrix_medium.py [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+样例
+Example 1:
+
+Input:
+[[0,1,2,0,0],
+ [1,0,0,2,1],
+ [0,1,0,0,0]]
+Output:
+2
+Example 2:
```
Comment (zjkang) [inline python/bfs_heap_topological_sort/bfs/lintcode_598_zombie_in_matrix_medium.py]: format

File: python/bfs_heap_topological_sort/bfs/lintcode_598_zombie_in_matrix_medium.py [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+        M = len(grid)
+        N = len(grid[0])
+        
+        queue = self.build_queue(grid, M, N)
+    
+        longest = self.get_longeset_time(queue, grid, M, N)
+        
+        if self.check_no_zeros(grid, M, N):
+            return longest
+        else:
```
Comment (zjkang) [inline python/bfs_heap_topological_sort/bfs/lintcode_598_zombie_in_matrix_medium.py]: no need to return extra `else:`, you can directly return -1

File: python/bfs_heap_topological_sort/bfs/lintcode_630_knight_shortest_path_ii_medium.py [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+                return path
+            
+            for d in directions:
+                nx, ny = x + d[0], y + d[1]
+                
+                if 0 <= nx < M and 0 <= ny < N and grid[nx][ny] == 0:
+                    grid[nx][ny] = 1
+                    queue.append((nx, ny, path + 1))
+        
+        if shortest == sys.maxsize:
```
Comment (zjkang) [inline python/bfs_heap_topological_sort/bfs/lintcode_630_knight_shortest_path_ii_medium.py:79]: only return -1?

File: python/bfs_heap_topological_sort/topo/lintcode_127_topological_sorting_medium.py [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+"""
+
+
+class Solution:
+    """
+    @param: graph: A list of Directed graph node
+    @return: Any topological order for the given graph.
+    """
+    def topSort(self, graph):
+        # write your code here
```
Comment (zjkang) [inline python/bfs_heap_topological_sort/topo/lintcode_127_topological_sorting_medium.py:41]: minor: remove comment

**Review by zjkang — CHANGES_REQUESTED** [review]
fix naming for file 'leetcode_210_course_schedule_ii_medium_frq5-.py'.

File: python/bfs_heap_topological_sort/bfs/lintcode_573_build_post_office_ii_hard.py (hunk 4 above)
Comment (wodeleeway) [reply python/bfs_heap_topological_sort/bfs/lintcode_573_build_post_office_ii_hard.py:27]: No, not being told. but we need to think fo 3 cases:
1) More houses, less empty spaces
2) More empty spaces, less houses
3) balanced houses and empty spaces

File: python/bfs_heap_topological_sort/bfs/lintcode_630_knight_shortest_path_ii_medium.py (hunk 12 above)
Comment (wodeleeway) [reply python/bfs_heap_topological_sort/bfs/lintcode_630_knight_shortest_path_ii_medium.py:79]: Yeah, there is path return inside the while loop

**Review by zjkang — APPROVED** [review]

**Pull request merged**